Store particles in vector container

gassmoeller opened the pull request:
Here are some ideas on how to implement #11206. This is not complete, but I will probably not get back to it in the next few weeks and I wanted to put it up here so that everyone interested can take a look. The new container is implemented and some basic functionality is working, but there are still bugs, and some functionality is missing completely (e.g. ghost particles, we may remove the ghost_particles container entirely, and simply determine if a certain cell is a ghost cell, in this case all particles in that cell are ghost particles).

blaisb commented:
> Here are some ideas on how to implement #11206. This is not complete, but I will probably not get back to it in the next few weeks and I wanted to put it up here so that everyone interested can take a look. The new container is implemented and some basic functionality is working, but there are still bugs, and some functionality is missing completely (e.g. ghost particles, we may remove the ghost_particles container entirely, and simply determine if a certain cell is a ghost cell, in this case all particles in that cell are ghost particles).

Very much looking forward to this. It seems like some bugs were introduced lately in particles (see #11591). Clearly we need additional tests for the particle library... :)

blaisb commented:
@gassmoeller I have given it a good look and I have not found any glaring issue. The output of the test changed because particles are printed in their order I guess? I will relook at the content soon to make sure.

However, I am a bit unsure about removing the existence of ghost particles. Please just keep in mind in the design that some of us are doing stuff that is just particle-particle. Consequently, for this we need to be able to quickly identify ghosts and non-ghosts in a context in which we are not necessarily going through the cells.  It would be very important for us to keep in mind that we often loop over particles outside of a loop over the cells :). For us this is critical.

gassmoeller commented:
> It would be very important for us to keep in mind that we often loop over particles outside of a loop over the cells :). For us this is critical.

Yes, that is one of the reasons I wanted to put this up early. Also so far I simply did not implement ghost particles in the new format (as in not at all), so I was curious to discuss in what form they are used. When you say you do not loop over cells, how do you loop over particles? I.e. does it matter that you can not specifically loop only over the ghost particles, but instead always loop over all particles (which accidentally happens in the order of containing cells)?

blaisb commented:
> > It would be very important for us to keep in mind that we often loop over particles outside of a loop over the cells :). For us this is critical.
> 
> Yes, that is one of the reasons I wanted to put this up early. Also so far I simply did not implement ghost particles in the new format (as in not at all), so I was curious to discuss in what form they are used. When you say you do not loop over cells, how do you loop over particles? I.e. does it matter that you can not specifically loop only over the ghost particles, but instead always loop over all particles (which accidentally happens in the order of containing cells)?

:) I think it's great to be able to communicate on this beforehand. I really appreciate it.
So we do three core tasks with the particles:
- Establish a contact list (this is achieved by looping over the mesh and then looping over the particles of each cell). Thus we access the particles in a cell->particle manner.
- Calculate the forces on the particle (this is done through the contact list). Thus we access the particles in a contact_list->particle manner. There we differentiate between regular and ghost (ghost particle don't need to "feel" the reciprocate of a contact).
- Integrate the position/velocity of the particle. There we do something like Verlet integration. In this case we loop directly over the particles because they already have the force they feel as a property. Additionnaly, we only integrate the motion of the particles that are not ghost. This part is critical for us since we do it every iteration. So there are two issues. We need to loop over all particles without accessing them through cells and we need to be able to loop over only regular particles. Right now the current structure allows us to do that.

gassmoeller commented:
Ok, I think this is much closer now. I re-enabled the ghost particle functionality and with some small changes all tests in deal.II  pass (some minor functionality still works differently than before). At this point this is close enough so that @blaisb, @peterrum  or @kronbichler could give it a try if you want. There are still performance problems, I suspect ParticleAccessor::get_surrounding_cell() may be slow for example, but others too. I will have to leave it at this stage for a few weeks, but feel free to comment and I will get back to it when I can.

blaisb commented:
> Ok, I think this is much closer now. I re-enabled the ghost particle functionality and with some small changes all tests in deal.II pass (some minor functionality still works differently than before). At this point this is close enough so that @blaisb, @peterrum or @kronbichler could give it a try if you want. There are still performance problems, I suspect ParticleAccessor::get_surrounding_cell() may be slow for example, but others too. I will have to leave it at this stage for a few weeks, but feel free to comment and I will get back to it when I can.

I will need to give a proper benchmark, but my initial tests show that it's about 1.8X slower than the previous implementation on our benchmarks. I think this is a great start considering that there is still room for improvement :). I'll try to help as much as possible with the PR.

gassmoeller commented:
> I will need to give a proper benchmark, but my initial tests show that it's about 1.8X slower than the previous implementation on our benchmarks.

Ha, I am happy it works at all :-). I had to make some small adjustments to the branch to make it work with ASPECT. I also found and fixed one major performance issue. The `update_cached_numbers()` function was optimized for the map container, but was really bad with the new container, because it called `get_surrounding_cell` for every particle. Give it a new try, now it is about as fast as the old container for me.

Does anyone know if there is a fast way to create an active_cell_iterator from an active cell index? Currently I do: `auto cell = triangulation.begin(); std::advance(cell, active_cell_index);` but that is not really efficient.

bangerth commented:
There is no way to do that from an `active_cell_index` because iterators are based on level/index-within-level pairs. There is no good way to translate between the two.

Do you need to use active_cell_index? Or could you use level/index instead? Even `CellId` would be faster.

blaisb commented:
> > I will need to give a proper benchmark, but my initial tests show that it's about 1.8X slower than the previous implementation on our benchmarks.
> 
> Ha, I am happy it works at all :-). I had to make some small adjustments to the branch to make it work with ASPECT. I also found and fixed one major performance issue. The `update_cached_numbers()` function was optimized for the map container, but was really bad with the new container, because it called `get_surrounding_cell` for every particle. Give it a new try, now it is about as fast as the old container for me.
> 
> Does anyone know if there is a fast way to create an active_cell_iterator from an active cell index? Currently I do: `auto cell = triangulation.begin(); std::advance(cell, active_cell_index);` but that is not really efficient.

Not only does it works straight out of the box, but now it is actually approximately as fast as the previous implementation. I am sure there is still some room for improvement, but this about as fast as the previous way it was done for what we do with it :)!

gassmoeller commented:
> Do you need to use active_cell_index? Or could you use level/index instead? Even CellId would be faster.

Maybe I could use level index instead. active_cell_index is the fastest way to index into the particle container that is why I chose it. Currently we also allow creating ParticleAccessors without access to a triangulation and I can not translate from level/index to active_cell_index without a triangulation. On the other hand maybe it makes sense to require a triangulation for particle iterators/accessors? After all the ParticleHandler that stores the particles should be connected to a triangulation anyway. This may break some tests and behavior though.

bangerth commented:
I think the question is what cases we want to cater to. deal.II is fundamentally a library for mesh-based finite element methods, and it would make sense to say that the particle implementation uses this fundamental premise as well. That means that every particle should have some sort of cell association. The question would then only be how that association is stored: Are particles keyed off cells (as they are currently), or is the cell simply stored as a property of the particle as I suspect @blaisb would prefer?

blaisb commented:
> I think the question is what cases we want to cater to. deal.II is fundamentally a library for mesh-based finite element methods, and it would make sense to say that the particle implementation uses this fundamental premise as well. That means that every particle should have some sort of cell association. The question would then only be how that association is stored: Are particles keyed off cells (as they are currently), or is the cell simply stored as a property of the particle as I suspect @blaisb would prefer?

That is a very good question.
I agree that deal.II is fundamentally a library for mesh-based finite elements methods and, as such, I think it is reasonable that particles should always "live" on a background mesh or grid. To me, this is a perfectly reasonable assumption.

I am not sure I mind if particles are keyed by cells or if cells are a property of them. In our case, we access particles both ways:
- Loop through cells to use the cell for contact detection
- Loop through particles for integration and other stuff.

I suspect most FEM code will use particles in a similar fashion. You will loop over the cell to couple the FEM and the particles, yet sometimes you will need to do particle integration or something similar, and this will be a loop over particles.
To me the important would be that accessing particles in both fashion remains easy to achieve. With the multimap, looping over particles was not difficult, but there were two issues:
- Particle data was spread out all over the memory;
- Getting a particle from it's ID was very difficult. You had to generate an auxiliary structure.

I am afraid I am unable at the moment to do the cost/benefit analysis of keying particles to cells, or having cells as property. People will always access the particles either directly, or through cells. I do not have your level of knowledge on the data structure used therein deal.II and what each choice would entail. Maybe this is something we could eventually have a meeting to discuss? Including @kronbichler   @peterrum  and @luca-heltai  could also be very useful. You are all more knowledgeable on this than me regrettably...

peterrum commented on include/deal.II/particles/particle_accessor.h:
```diff
@@ -254,24 +272,24 @@ namespace Particles
 
     /**
      * An iterator into the container of particles. Obviously,
      * this accessor is invalidated if the container changes.
      */
-    typename std::multimap<internal::LevelInd,
-                           Particle<dim, spacedim>>::iterator particle;
+    unsigned int active_cell_index;
+
+    unsigned int particle_index;
```
The comment needs to be updated.

peterrum commented on include/deal.II/particles/particle_accessor.h:
```diff
@@ -445,10 +506,13 @@ namespace Particles
 
-    const typename Triangulation<dim, spacedim>::cell_iterator cell(
-      &triangulation, particle->first.first, particle->first.second);
+    typename Triangulation<dim, spacedim>::active_cell_iterator cell =
+      triangulation.begin_active();
+    std::advance(cell, active_cell_index);
```
I guess we should cache in `Triangulation` a translation between active_cell_index and a level/index pair.

peterrum commented on include/deal.II/particles/particle_handler.h:
```diff
@@ -1020,7 +1034,11 @@ namespace Particles
   ParticleHandler<dim, spacedim>::begin()
   {
-    return particle_iterator(particles, particles.begin());
+    for (unsigned int i = 0; i < particles.size(); ++i)
+      if (particles[i].size() != 0)
```
Shouldn't this info  be cached?

peterrum commented on source/particles/particle_handler.cc:
```diff
@@ -347,7 +355,59 @@ namespace Particles
   {
-    particles.erase(particle->particle);
+    const unsigned int active_cell_index =
+      particle->get_surrounding_cell(*triangulation)->active_cell_index();
```
```suggestion
      particle->active_cell_index;
```

bangerth replied:
ha, yes :-)

peterrum commented on source/particles/particle_handler.cc:
```diff
@@ -358,18 +418,34 @@ namespace Particles
+    if (particles.size() == 0)
+      particles.resize(triangulation->n_active_cells());
+
+    typename Triangulation<dim, spacedim>::active_cell_iterator cell_to_insert;
```
Name `cell_to_insert` is a bit misleading. The particle is inserted.

peterrum commented on source/particles/particle_handler.cc:
```diff
@@ -358,18 +418,34 @@ namespace Particles
+      cell_to_insert = cell;
+    else
+      {
+        cell_to_insert = triangulation->begin_active();
+        while (cell_to_insert->is_locally_owned() == false)
+          ++cell_to_insert;
+
+        Assert(
+          cell_to_insert->is_locally_owned(),
+          ExcMessage(
+            "ParticleHandler<dim, spacedim>::insert_particle can only insert particles into locally owned cells, but none was found."));
```
The particle is inserted into any locally-owned cell?

peterrum commented on source/particles/particle_handler.cc:
```diff
@@ -358,18 +418,34 @@ namespace Particles
+      }
+
+    const unsigned int active_cell_index = cell_to_insert->active_cell_index();
+    particles[active_cell_index].push_back(particle);
```
Maybe we could think about sorting these vectors in the future? About how many particles in a cell are we talking about in general?

gassmoeller replied:
I am not sure how you would sort them here. What would be the sorting criteria? The only useful criteria I could see is the order of property pool handles, so that all particle data is stored continuously in the order of particle iteration. However, that is a sort on all local particles, I would not want to do that every time a single particle is inserted/removed/moves to a different cell. Currently this happens during every mesh refinement/repartition.

peterrum replied:
Maybe once globally?

kronbichler replied:
I agree, sorting things upon mesh refinement sounds reasonable unless we have indication about that with cache analysis data (because it is mostly memory fragmentation with poor prefetching I could imagine as a potential problem here).

peterrum commented on source/particles/particle_handler.cc:
```diff
@@ -1033,7 +1110,7 @@ namespace Particles
 
           const unsigned int closest_vertex =
             GridTools::find_closest_vertex_of_cell<dim, spacedim>(
-              current_cell, out_particle->get_location(), *mapping);
+              current_cell, out_particle->get_location());
```
?

peterrum commented on source/particles/particle_handler.cc:
```diff
@@ -1208,7 +1274,7 @@ namespace Particles
         if (cell->is_locally_owned())
           {
             std::set<unsigned int> cell_to_neighbor_subdomain;
-            for (const unsigned int v : cell->vertex_indices())
+            for (const unsigned int v : GeometryInfo<dim>::vertex_indices())
```
I think some of your changes revert the changes made during the simplex workshop to make the `ParticleHandler` work on simplex meshes. Could you check these instances?

peterrum commented on source/particles/particle_handler.cc:
```diff
@@ -1533,21 +1604,23 @@ namespace Particles
         recv_data_it = static_cast<const char *>(recv_data_it) + cellid_size;
 
         const typename Triangulation<dim, spacedim>::active_cell_iterator cell =
-          triangulation->create_cell_iterator(id);
+          id.to_cell(*triangulation);
```
This function has been deprecated. I guess this is accidental revert of the change?

peterrum commented on source/particles/particle_handler.cc:
```diff
@@ -1799,68 +1874,32 @@ namespace Particles
 
     switch (status)
       {
-        case parallel::TriangulationBase<dim, spacedim>::CELL_PERSIST:
-        case parallel::TriangulationBase<dim, spacedim>::CELL_REFINE:
+        case parallel::distributed::Triangulation<dim, spacedim>::CELL_PERSIST:
+        case parallel::distributed::Triangulation<dim, spacedim>::CELL_REFINE:
```
?

peterrum commented on source/particles/particle_handler.cc:
```diff
@@ -1892,76 +1931,42 @@ namespace Particles
 
     switch (status)
       {
-        case parallel::TriangulationBase<dim, spacedim>::CELL_PERSIST:
+        case parallel::distributed::Triangulation<dim, spacedim>::CELL_PERSIST:
```
?

peterrum reviewed (commented):
I like the direction very much! A general question if we want to sort particles and if yes when? 

As far as I see it in the long run, the particles might not need to store `Particle` objects but only indices. Furthermore, we have at the moment the classes `Particle` -> `ParticleAccessor` -> `ParticleIterator`, which feels a layer of indirection too much (if we compare it with `Triangulation` and `DoFHandler`).

> Getting a particle from it's ID was very difficult. You had to generate an auxiliary structure.

@blaisb  I think we could set up the aux data structures within the `ParticleHandler` as a cache (particle index -> active cell index + local particle index).

blaisb commented:
> I like the direction very much! A general question if we want to sort particles and if yes when?
> 
> As far as I see it in the long run, the particles might not need to store `Particle` objects but only indices. Furthermore, we have at the moment the classes `Particle` -> `ParticleAccessor` -> `ParticleIterator`, which feels a layer of indirection too much (if we compare it with `Triangulation` and `DoFHandler`).
> 
> > Getting a particle from it's ID was very difficult. You had to generate an auxiliary structure.

Agreed. This would move us away from the Vector of class mindset to just a class of vector containers. I think it would be more efficient. At the moment I see little to no reason to keep a Particle class.

> 
> @blaisb I think we could set up the aux data structures within the `ParticleHandler` as a cache (particle index -> active cell index + local particle index).

Agreed :)! I'll be glad to help writing that.

bangerth commented on source/particles/particle_handler.cc:
```diff
@@ -223,34 +223,32 @@ namespace Particles
-      triangulation->begin_active();
 
-    for (const auto &particle : *this)
+    for (const std::vector<Particle<dim,spacedim>> &particles_in_cell : particles)
```
```suggestion
    for (const auto&particles_in_cell : particles)
```
maybe?

bangerth commented on source/particles/particle_handler.cc:
```diff
@@ -314,22 +315,29 @@ namespace Particles
+        else
+          {
+            auto end =
+              particle_iterator(container,
```
make it easier to read: if you declare an `end` variable, also declare a `begin` variable. or define both in-line as you currently do for the begin iterator

bangerth commented on source/particles/particle_handler.cc:
```diff
@@ -314,22 +315,29 @@ namespace Particles
+              particle_iterator(container,
+                                active_cell_index,
+                                container[active_cell_index].size() - 1);
+            end++;
```
```suggestion
            ++end;
```

bangerth reviewed (commented):
I looked through this, and I'm not so sure yet whether this is a direction I like. The `std::vector<std::vector<...>>` is an expensive data structure if you don't have a lot of particles on *all* cells.

Have you thought about alternatives? I'm not sure what precisely we want. I guess these are the conditions:
* Random insertion and removal
* Linear access
* The ability to find the particles located on a cell
* The ability to find a cell based on a particular
* Be memory efficient for cases where there are many and where there are few particles per cell (including cases with many cells with no particles at all).
Is that about it?

The `std::multimap` was one way to achieve this, but then you measured that iteration takes a long time. I wonder whether `std::unordered_multimap` might be a better choice? Or something akin to a CSR storage for sparse matrices?

I haven't read through the `particle_handler.cc` file in detail yet, but thought we'd talk about the data structure first.

bangerth commented:
Or maybe `std::map<std::vector>` would be another choice. That would address the issue that for some cells, we may not have any particles, and if the accessor stored which vector it is currently operating on, the outer map lookup would be amortized.

bangerth commented:
@gassmoeller and I just talked and the consensus was that we should postpone this till after the release.

blaisb commented:
> @gassmoeller and I just talked and the consensus was that we should postpone this till after the release.

Good idea. It's a complex topic with multiple implications. Let's not rush into things, it's not like the current structure is dramatically bad either...

gassmoeller commented:
I have rebased this PR to the recent master and fixed some incompatibilities between the branch and master. All particle tests pass now. @kronbichler and @peterrum I have not yet addressed all of your comments on this PR, but if you want to give this a try in terms of performance or looking for bugs feel free to test it. My main goal for next week is to get this into shape to be merged. Before that I want to run some scaling tests with it to make sure we do not unintentionally break large-scale performance.

peterrum commented:
> I have rebased this PR to the recent master and fixed some incompatibilities between the branch and master. All particle tests pass now.

Thanks! I will wait for a next round of review until you have addressed the old comments. There are still a bunch of cases where you have overridden changes made in the last months on master.

kronbichler commented:
I ran a short test, a variant of step-68 but run in 3D and with polynomial degree 4 (this was designed to be a stress test for `FEPointEvaluation`, so don't wonder about the degree). In serial with 10 time steps and 1.5 million particles, I get
```
                               master    this PR
create particles               10.29 s   10.21 s
Euler step interpolated        1.801 s   1.427 s
Sort particles into subdomains 2.726 s   1.903 s
```
The more interesting thing is that the evaluation of the velocity field takes 0.12 seconds per time step, whereas the manipulations of the particles take now 0.03 seconds per time step. On master, the timing was 0.07 seconds per time step, hence, explaining the difference. Just for completeness, if I go away from the high degree of 4 to a polynomial degree of 1 instead the evaluation of the velocity field takes 0.03 seconds per time step, so now we finally can compare the time to access the particles with the access of the background velocity.

On 4 MPI ranks and after 30 time steps, I see
```
                               master    this PR
create particles               4.101 s   4.125 s
Euler step interpolated        2.050 s   1.659 s
Sort particles into subdomains 3.902 s   2.517 s
```
Also the sorting of particles is much faster now, so I would consider this direction a very good step forward. Thanks @gassmoeller !

masterleinad commented on source/particles/particle_handler.cc:
```diff
@@ -456,22 +554,20 @@ namespace Particles
-      particles.find(std::make_pair(cells[0]->level(), cells[0]->index()));
     for (unsigned int i = 0; i < cells.size(); ++i)
       {
         internal::LevelInd current_cell(cells[i]->level(), cells[i]->index());
```
```suggestion
```

masterleinad commented on source/particles/particle_handler.cc:
```diff
@@ -938,53 +1034,56 @@ namespace Particles
+        real_locations.reserve(n_pic);
+        reference_locations.resize(n_pic);
+
+        for (const auto particle : pic)
```
```suggestion
        for (const auto& particle : pic)
```

blaisb commented:
> I ran a short test, a variant of step-68 but run in 3D and with polynomial degree 4 (this was designed to be a stress test for `FEPointEvaluation`, so don't wonder about the degree). In serial with 10 time steps and 1.5 million particles, I get
> 
> ```
>                                master    this PR
> create particles               10.29 s   10.21 s
> Euler step interpolated        1.801 s   1.427 s
> Sort particles into subdomains 2.726 s   1.903 s
> ```
> 
> The more interesting thing is that the evaluation of the velocity field takes 0.12 seconds per time step, whereas the manipulations of the particles take now 0.03 seconds per time step. On master, the timing was 0.07 seconds per time step, hence, explaining the difference. Just for completeness, if I go away from the high degree of 4 to a polynomial degree of 1 instead the evaluation of the velocity field takes 0.03 seconds per time step, so now we finally can compare the time to access the particles with the access of the background velocity.
> 
> On 4 MPI ranks and after 30 time steps, I see
> 
> ```
>                                master    this PR
> create particles               4.101 s   4.125 s
> Euler step interpolated        2.050 s   1.659 s
> Sort particles into subdomains 3.902 s   2.517 s
> ```
> 
> Also the sorting of particles is much faster now, so I would consider this direction a very good step forward. Thanks @gassmoeller !

Wow these are significant changes! Sorting the particles becomes close to 2x faster...

gassmoeller commented:
All right, I think I addressed all comments and made this as ready as a change of the data structure can be. I am looking for another round of review or comments, but please focus on getting this into shape for merging first (so only bugfixes in this PR) and postpone other suggestions for new features and optional improvements to follow up PRs.

This already promises a quite substantial performance benefit as Martin mentioned. An example from ASPECT with a significant number of particles (2.4e7 particles on 32 processes):

deal.II 9.3:

```
| Particles: Advect                |        22 |      7.59s |        17% |
| Particles: Copy                  |        11 |      3.54s |       7.9% |
| Particles: Generate              |         1 |     0.953s |       2.1% |
| Particles: Initialization        |         1 |  0.000334s |         0% |
| Particles: Initialize properties |         1 |      0.26s |      0.58% |
| Particles: Interpolate           |        11 |     0.888s |         2% |
| Particles: Sort                  |        22 |      4.73s |        11% |
| Refine mesh structure, part 1    |        11 |      10.6s |        24% |
| Refine mesh structure, part 2    |        11 |      6.95s |        16% |
```

This branch:

```
| Particles: Advect                |        22 |      6.65s |        21% |
| Particles: Copy                  |        11 |      1.54s |       4.9% |
| Particles: Generate              |         1 |     0.874s |       2.8% |
| Particles: Initialization        |         1 |  0.000325s |         0% |
| Particles: Initialize properties |         1 |     0.258s |      0.83% |
| Particles: Interpolate           |        11 |     0.554s |       1.8% |
| Particles: Sort                  |        22 |      1.41s |       4.5% |
| Refine mesh structure, part 1    |        11 |      8.24s |        26% |
| Refine mesh structure, part 2    |        11 |      2.97s |       9.5% |
```

There are at least two ideas for future PRs that should further improve performance:
- The accessor currently stores a cell iterator and a cell index. This is duplicative and may make iteration slower, but the cell index is used heavily for accessing particles so I am not sure if using `cell->active_cell_index()` would slow things down. Will investigate in a separate branch.
- As we discussed moving a lot of the code of the Particle class into the accessor will remove an unnecessary layer of indirection, and also reduce the memory requirement, because we only store a single pointer to the property pool, not one per particle. This will also make moving/copying particles faster.

blaisb commented:
> All right, I think I addressed all comments and made this as ready as a change of the data structure can be. I am looking for another round of review or comments, but please focus on getting this into shape for merging first (so only bugfixes in this PR) and postpone other suggestions for new features and optional improvements to follow up PRs.
> 
> This already promises a quite substantial performance benefit as Martin mentioned. An example from ASPECT with a significant number of particles (2.4e7 particles on 32 processes):
> 
> deal.II 9.3:
> 
> ```
> | Particles: Advect                |        22 |      7.59s |        17% |
> | Particles: Copy                  |        11 |      3.54s |       7.9% |
> | Particles: Generate              |         1 |     0.953s |       2.1% |
> | Particles: Initialization        |         1 |  0.000334s |         0% |
> | Particles: Initialize properties |         1 |      0.26s |      0.58% |
> | Particles: Interpolate           |        11 |     0.888s |         2% |
> | Particles: Sort                  |        22 |      4.73s |        11% |
> | Refine mesh structure, part 1    |        11 |      10.6s |        24% |
> | Refine mesh structure, part 2    |        11 |      6.95s |        16% |
> ```
> 
> This branch:
> 
> ```
> | Particles: Advect                |        22 |      6.65s |        21% |
> | Particles: Copy                  |        11 |      1.54s |       4.9% |
> | Particles: Generate              |         1 |     0.874s |       2.8% |
> | Particles: Initialization        |         1 |  0.000325s |         0% |
> | Particles: Initialize properties |         1 |     0.258s |      0.83% |
> | Particles: Interpolate           |        11 |     0.554s |       1.8% |
> | Particles: Sort                  |        22 |      1.41s |       4.5% |
> | Refine mesh structure, part 1    |        11 |      8.24s |        26% |
> | Refine mesh structure, part 2    |        11 |      2.97s |       9.5% |
> ```
> 
> There are at least two ideas for future PRs that should further improve performance:
> 
> * The accessor currently stores a cell iterator and a cell index. This is duplicative and may make iteration slower, but the cell index is used heavily for accessing particles so I am not sure if using `cell->active_cell_index()` would slow things down. Will investigate in a separate branch.
> * As we discussed moving a lot of the code of the Particle class into the accessor will remove an unnecessary layer of indirection, and also reduce the memory requirement, because we only store a single pointer to the property pool, not one per particle. This will also make moving/copying particles faster.

Looking good from my POV. I agree with you that this PR is already a lot of content and that future improvement should be done in other PRs.

peterrum commented on source/particles/particle_handler.cc:
```diff
@@ -283,13 +296,16 @@ namespace Particles
-      return particles.count(found_cell);
-    else if (cell->is_ghost())
-      return ghost_particles.count(found_cell);
+    if (cell->is_artificial() == false)
+      {
+        if (cell->is_locally_owned() == true)
+          return particles[cell->active_cell_index()].size();
+        else
+          return ghost_particles[cell->active_cell_index()].size();
+      }
```
If we'd merge the vectors:
```suggestion
    Assert(cell->is_artificial() == false, ExcInternalError());
    return particles[cell->active_cell_index()].size();
```

peterrum commented on source/particles/particle_handler.cc:
```diff
@@ -351,7 +376,70 @@ namespace Particles
     const ParticleHandler<dim, spacedim>::particle_iterator &particle)
   {
-    particles.erase(particle->particle);
+    const unsigned int active_cell_index = particle->active_cell_index;
+
+    if (particles[active_cell_index].size() > 1)
```
Shouldn't we check here that the cell is not ghost?

gassmoeller replied:
You are right. Even better, since the accessor stores the point to the container it acts upon we should use that directly. I guess nobody tried removing a ghost particle individually so far (it is not something you would usually do).

peterrum commented on source/particles/particle_handler.cc:
```diff
@@ -939,52 +1030,65 @@ namespace Particles
     std::vector<particle_iterator> particles_out_of_cell;
-    particles_out_of_cell.reserve(n_locally_owned_particles());
+
+    particles_out_of_cell.reserve(n_locally_owned_particles() / 4);
```
Divided by 4?

gassmoeller replied:
Just a guess for how many particles need sorting into a different cell. I added documentation.

gassmoeller commented on include/deal.II/particles/particle_handler.h:
```diff
@@ -833,21 +850,26 @@ namespace Particles
      * Set of particles currently living in the local domain, organized by
      * the level/index of the cell they are in.
      */
-    std::multimap<internal::LevelInd, Particle<dim, spacedim>> particles;
+    particle_container particles;
 
     /**
-     * Set of particles that currently live in the ghost cells of the local
-     * domain, organized by the level/index of the cell they are in. These
-     * particles are equivalent to the ghost entries in distributed vectors.
+     * Set of particles currently living in the local domain, organized by
+     * the level/index of the cell they are in.
      */
-    std::multimap<internal::LevelInd, Particle<dim, spacedim>> ghost_particles;
+    particle_container ghost_particles;
```
I agree, but it is not easy to merge them right now. For one we allow inserting particles into the main particles container even into ghost cells, these partices are then sorted and communicated in sort_particles_into_cells, while particles in the ghost_particles container are never sent. Additionally some functions only work on the one container, others on the other. I would like to chat about how to unify them during the workshop, also to reduce the memory overhead, but I would like to do it in a separate PR.

peterrum replied:
Sure. This was more a question. We will have enough time next week :)

gassmoeller commented on include/deal.II/particles/particle_accessor.h:
```diff
@@ -239,36 +254,65 @@ namespace Particles
+     * read from the member variable cell, but is used in many performance
+     * critical functions and is therefore stored and updated separately.
+     */
+    unsigned int active_cell_index;
```
That wont work, the index changes whenever the cell changes.

gassmoeller commented on include/deal.II/particles/particle_accessor.h:
```diff
@@ -286,29 +330,38 @@ namespace Particles
+    const typename Triangulation<dim, spacedim>::active_cell_iterator &cell,
+    const unsigned int particle_index_within_cell)
+    : particles(const_cast<particle_container *>(&particles))
+    , cell(cell)
+    , particle_index_within_cell(particle_index_within_cell)
```
Hm, I see that you want to place it into the initializer list, but I find this terribly confusing to read. Is there a reason to not have the if condition below?

peterrum replied:
If you cannot make it const not... So let's keep it!

gassmoeller commented:
I think I addressed all your comments.

peterrum reviewed (approved):
What is missing from my side is a change-log entry. Else I am all for merging!

peterrum commented:
/rebuild

kronbichler commented:
I suggest we merge this tomorrow before the workshop starts, unless someone has strong reasons to postpone. We might tackle some of the points @gassmoeller mentions in the next days.

blaisb commented:
> I suggest we merge this tomorrow before the workshop starts, unless someone has strong reasons to postpone. We might tackle some of the points @gassmoeller mentions in the next days.

Agreed. I'd like to tackle one of these myself.

gassmoeller commented:
I added a changelog entry. I put it into major for now, because I thought we could collect all the improvements of next week in one entry (and also add all of you as authors then of course).
